File: iosApp/iosApp/Helper/PokemonNavigator.swift
//
// Created by Alif Phincon on 29/09/25.
//

import Combine
import ComposeApp
import Foundation

final class PokemonNavigatorCombine: ObservableObject {

   private let navigator: PokemonNavigator
   let currentSubject: CurrentValueSubject<Pokedex?, Never>

   init(results: [Pokedex], startIndex: Int = 0) {
       self.navigator = PokemonNavigator(allResults: results, startIndex: Int32(startIndex))
       self.currentSubject = CurrentValueSubject(self.navigator.current)
   }

   func moveNext() {
       let new = navigator.moveNext()
       currentSubject.send(new)
   }

   func movePrevious() {
       let new = navigator.movePrevious()
       currentSubject.send(new)
   }
}

File: composeApp/src/commonMain/kotlin/com/appwork/pokeapi_kmm/data/PokemonNavigator.kt
package com.appwork.pokeapi_kmm.data

import com.appwork.pokeapi_kmm.models.Pokedex

class PokemonNavigator(
    private val allResults: List<Pokedex>,
    startIndex: Int? = 0
) {
    private var currentIndex: Int? = if (startIndex in allResults.indices) startIndex else null

    val current: Pokedex?
        get() = currentIndex?.let { allResults[it] }

    fun moveNext(): Pokedex? {
        val index = currentIndex
        if (index != null && index + 1 < allResults.size) {
            currentIndex = index + 1
        }
        return current
    }

    fun movePrevious(): Pokedex? {
        val index = currentIndex
        if (index != null && index - 1 >= 0) {
            currentIndex = index - 1
        }
        return current
    }
}


